### PR TITLE
assert convertibility and typing with context

### PR DESCRIPTION
**Pull request by Rehan-MALAK**
@fblanqui @elhaddadyacine please make comments about the syntax you are expecting here  #298 and @fblanqui maybe what you had in mind concerning @gabrielhdt #309

please note that currently, it allows to do

```
assert {(n m:N) (o:O)} f1 n m o ≡ f2 n m o
```
but not
```
assert {n m o} f1 n m o ≡ f2 n m o
```

### CODE REVIEW

**Comment by fblanqui** [comment]
For the syntax, I propose:
(assert|assertnot) decl1, .., decln |- a = b
decl := x [: A] [:= t]
(using unicode symbols for |-, = and := )
types and definitions should be optional

File: src/core/legacy_lexer.mll [hunk 1]
```diff
@@ -24,7 +24,7 @@ let locate_lexbuf : Lexing.lexbuf -> Pos.pos = fun lexbuf ->
 
 type token =
   (* Special characters and symbols. *)
-  | L_SQB | R_SQB | L_PAR | R_PAR | ARROW | LARROW | FARROW | DEFEQ | COMMA
+  | L_SQB | R_SQB | L_PAR | R_PAR | LEFTBRA | RIGHTBRA | ARROW | LARROW | FARROW | DEFEQ | COMMA
```
Comment (fblanqui) [inline src/core/legacy_lexer.mll]: Why do you change the legacy parser? The legacy parser is a parser for the current version of Dedukti. It should not include things that are not in Dedukti.

**Comment by fblanqui** [comment]
Thanks to update syntax.bnf as well.

File: src/core/legacy_lexer.mll (hunk 1 above)
Comment (fblanqui) [reply src/core/legacy_lexer.mll]: Well, in this case, this is harmless. So, that's fine.

**Comment by rlepigre** [comment]
> For the syntax, I propose:
> (assert|assertnot) decl1, .., decln |- a = b
> decl := x [: A] [:= t]
> (using unicode symbols for |-, = and := )
> types and definitions should be optional

Yes, I agree with @fblanqui that is a much better syntax.

I thing that I don't agree with in what you have done is the extension of the legacy syntax. Only the standard syntax should be extended, the legacy syntax is only here for compatibility.

**Comment by Rehan-MALAK** [comment]
I added initially the extension of the legacy syntax because this feature was actually used by @barras for some time. It's not that important now.

But if we really care more about compatibility than trying to limit divergences between old and new syntax, there are several things to change in legacy parser. I will add a quick test in #315.

**Comment by Rehan-MALAK** [comment]
> use the argument syntax in the context

yes, the idea was to use what's already done : the fact that arg can be ``x`` or ``(x : X)`` or ``x y (z:X)``
with minimal modifications in core/src/parser.ml as this parser is going to go away anyway in N days

> more general contexts introduced by #309

yes, it's not finished, you guys can merge it now or consider it as a WIP and wait for #309

**Comment by Rehan-MALAK** [comment]
sorry  , just corrected a typo and rebased on current master

**Comment by fblanqui** [comment]
Thanks Rehan!

**Comment by Rehan-MALAK** [comment]
It seems that without write access, it's not possible to trigger a CI build (unless I close/reopen PR or push force another time)
https://stackoverflow.com/questions/17606874/trigger-a-travis-ci-rebuild-without-pushing-a-commit

**Comment by Rehan-MALAK** [comment]
the commit message is horrible ... can we do something about it, rewrite the history now before it's too late ?

**Comment by rlepigre** [comment]
I think we should not try and extend the mechanism introduced by this PR to contexts containing definitions yet (i.e., in #309). As @Rehan-MALAK noted, this will interact with the parser.

**Comment by fblanqui** [comment]
Don't worry. It is not so important.

**Comment by rlepigre** [comment]
> the commit message is horrible ... can we do something about it, rewrite the history now before it's too late ?

No, that's fine. We have had worse commit messages before...

**Comment by fblanqui** [comment]
But I don't understand the right access you are talking about.

**Comment by Rehan-MALAK** [comment]
well it's really confusing... 

    assert convertibility and typing with context (#320)
    
    * assert convertibility and typing with context
    
      * old syntax :
    
        #ASSERT {(n:N) (m:N) (o:O)} f1 n m o == f2 n m o.
        #ASSERT {(n:N) (m:N) (o:O)} f n m o : P.
    
      * new syntax :
    
        assert {(n m:N) (o:O)} f1 n m o ≡ f2 n m o
        assert {(n m:N) (o:O)} f n m o : P
    
    * no old syntax after all
    
    * introduce unicode ⊢
    
      assert n m (o:N) ⊢ f n m o ≡ f n m o
      assert n (m:N) ⊢ f n m : N ⇒ N

**Comment by fblanqui** [comment]
Ok, we can wait.

Le 26/03/2020 à 14:47, Rodolphe Lepigre a écrit :

**Pull request merged**